Implementation/57516 projects should be removable from list with the help of a menu item

akabiru opened the pull request:
# Ticket
<!-- Provide the link to respective work package -->

https://community.openproject.org/work_packages/57516

# What are you trying to accomplish?
<!-- Provide a description of the changes. -->

A project should be removable from a custom field on the project list via a menu action item

## Screenshots
<!-- Provide before/after screenshots, videos, or graphs for any visual changes; otherwise, remove this section -->

![Kapture 2024-09-13 at 13 38 48](https://github.com/user-attachments/assets/018eb248-a75a-4fbe-a8d8-716dc9dbe5db)

# What approach did you choose and why?

<!-- This section is a place for you to describe your thought process in making these changes.
     List any tradeoffs you made to take on or pay down tech debt.
     Describe any alternative approaches you considered and why you discarded them. -->

> [!IMPORTANT]  
> Adding a primary key to an existing table has a seemingly higher opportunity cost compared to the
alternative; composite primary keys which will have the carrying cost of introducing different data structure-
which have a higher chance of introducing conflicts or requiring additional handling (ergo- _carrying cost_
>
> In the migration- defining a PRIMARY KEY via Rails migration ensures an autoincremting column is defined as a BIGSERIAL type and existing records are backfilled automatically
>

akabiru committed:
[OP#57516] Projects should be removable from list with the help of a menu item

https://community.openproject.org/work_packages/57516

akabiru committed:
tests[Op#57516]: specify `delete_all` usage in custom fields projects delete service

As the `custom_fields_projects` table was not set up conventionally with a primary key, it is not possible to
leverage ActiveRecord#destroy as that's coupled to an id. We therefore use a direct SQL DELETE via `ActiveRecord#delete_all`
which has the disadvantage of skipping any callbacks- including `dependent:` relations.

No `dependent:` rules are strictly relevant in this context (at this time) but should that change, they would either need to be triggered manually
OR this table updated to Rails convention: "avoiding configuration"

akabiru committed:
tests[Op#57516]: revisit page to ensure new project is available

akabiru committed:
chore[Op#57516]: extract method, not found turbo stream response

akabiru commented on app/services/custom_fields/custom_field_projects/delete_service.rb:
```diff
@@ -0,0 +1,53 @@
+        ActiveRecord::Type::Boolean.new.cast(delete_result)
+      end
+
+      # `custom_fields_projects` table has no `id` column, hence no primary key. #destroy method would not work
+      # Note: `delete_all` goes straight to the database and does not trigger callbacks
+      #
+      # @return [Integer] number of rows deleted
+      def delete(custom_field_id:, project_id:)
+        CustomFieldsProject.transaction do
+          CustomFieldsProject.where(custom_field_id:, project_id:).delete_all
+        end
+      end
```
\cc @ulferts any thoughts on this approach?

ulferts replied:
I think it is ok to do this given what the data structure currently offers. 

But there might be a better way of doing it. I don't have first hand experience with it yet, though, so I don't particularly know the limitations. Since [Rails 7.1.2 composite primary keys are supported](https://blog.saeloun.com/2024/08/06/active-record-composite-primary-key/). I still don't quite understand how much of [this](https://guides.rubyonrails.org/active_record_composite_primary_keys.html) is only available in edge and how much is already in 7.1.2. so you'd have to look that up for yourself. 

But my quick test where I changed the `CustomFieldsProject` to include the specification for the composite primary key

```
class CustomFieldsProject < ApplicationRecord
  self.primary_key = %i[project_id custom_field_id]

  belongs_to :project
  belongs_to :custom_field
end
```

Let me to already do this in the console:

```
[3] pry(main)> CustomFieldsProject.find([992, 152]).destroy
  CustomFieldsProject Load (0.4ms)  SELECT "custom_fields_projects".* FROM "custom_fields_projects" WHERE "custom_fields_projects"."project_id" = $1 AND "custom_fields_projects"."custom_field_id" = $2 LIMIT $3  [["project_id", 992], ["custom_field_id", 152], ["LIMIT", 1]]
  TRANSACTION (0.1ms)  BEGIN
  CustomFieldsProject Destroy (0.7ms)  DELETE FROM "custom_fields_projects" WHERE "custom_fields_projects"."project_id" = $1 AND "custom_fields_projects"."custom_field_id" = $2  [["project_id", 992], ["custom_field_id", 152]]
  TRANSACTION (2.2ms)  COMMIT
=> #<CustomFieldsProject:0x0000000332b9e620 custom_field_id: 152, project_id: 992>
```

I can now also query for e.g. `CustomFieldsProject.last`. It still isn't a proper composite key in the database but it might well be good enough.

If that doesn't work, it might still be worthwhile to add an id column to the table and work through the necessary changes. It will probably make a whole lot of code simpler (e.g. in the controller). But I'd understand if you don't want to do that now.

akabiru replied:
Thank you! In the end adding an ID column seems like the better solution, albeit with higher opportunity cost compared to the composite primary keys which will have the carrying cost of a different data structure- which I have already hit issues with in one part of the application (ng auto-completer)

ulferts commented on app/controllers/admin/custom_fields/custom_field_projects_controller.rb:
```diff
@@ -110,11 +129,14 @@ def find_projects_to_activate_for_custom_field
 
-    respond_with_turbo_streams
+  def find_custom_field_project_to_destroy
+    @project = Project.find(params.to_unsafe_h[:custom_fields_project][:project_id])
```
I am rather sure that you don't need the `to_unsafe_h` here. You could just write

```
@project = Project.find(params[:custom_fields_project][:project_id])
```

ulferts replied:
Come to think of it, I don't understand why the project is fetched at all. Couldn't it also be written like this:

```
@custom_field_project = CustomFieldsProject.find_by!(custom_field: @custom_field, 
                                                     project: params[:custom_fields_project][:project_id])
```

That of course is only true if `@project` is really not used at all. But you'd safe a database request if it needn't be loaded.

akabiru replied:
Great catch, missed that! 👍🏾

ulferts commented on app/components/admin/custom_fields/custom_field_projects/row_component.rb:
```diff
@@ -37,8 +37,31 @@ def wrapper_uniq_by
+        private
+
+        def more_menu_detach_project
+          if User.current.admin && project.active?
```
You could also query for `User.current.allowed_to?(:select_custom_fields, project)`. This is more specific than the admin check (only admins will visit that page anyway), but allows to later on open up the page to non admins more easily. At the same time, it will check for the project being active as no permissions are granted on archived projects.

akabiru replied:
Good call! 👍🏾

ulferts reviewed (commented):
Hey @akabiru, 

I did not do a proper review as I only browsed through the code and didn't check out the functionality itself. I have a proposal on how you could add a composite primary key and by that work around the missing id column. Other than that, I added smaller remarks.

akabiru committed:
fix[Op#57516]: use more specific perm check

This is more specific than the admin check and allows easier extension to non-admins as long they
have the permission.

https://github.com/opf/openproject/pull/16694/files#r1759090033

akabiru committed:
chore[Op#57516]: remove unneded unsafe hash and extra DB Call

https://github.com/opf/openproject/pull/16694/files#r1759047117

akabiru committed:
fix[Op#57516]: add missing flash component stream

akabiru committed:
fix[Op#57516]: add missing primary key to custom fields projects

Adding a primary key to an existing table has a seemingly higher opportunity cost compared to the
alternative; composite primary keys which will have the carrying cost of introducing different data structure-
which have a higher chance of introducing conflicts or requiring additional handling (ergo- _carrying cost_

In the migration- defining a PRIMARY KEY via Rails migration ensures an autoincremting column is defined as a BIGSERIAL type
and existing records are backfilled via `row_number()` pg window function

akabiru commented on db/migrate/20240917105829_add_primary_key_to_custom_fields_projects.rb:
```diff
@@ -0,0 +1,50 @@
+      end
+    end
+  end
+end
```
> [!IMPORTANT]  
> Adding a primary key to an existing table has a seemingly higher opportunity cost compared to the
alternative; [composite primary keys](https://guides.rubyonrails.org/active_record_composite_primary_keys.html) which will have the carrying cost of introducing different data structure- which have a higher chance of introducing conflicts or requiring additional handling (ergo- _carrying cost_
>
> In the migration- defining a PRIMARY KEY via Rails migration ensures an auto-incrementing column is defined as a _BIGSERIAL_ type with the default value set to `nextval('custom_fields_projects_id_seq'::regclass)` and existing records are backfilled via `row_number()` pg window function
> 

---

```ruby
Migrating to AddPrimaryKeyToCustomFieldsProjects (20240917105829)
== 20240917105829 AddPrimaryKeyToCustomFieldsProjects: migrating ==============
-- add_column(:custom_fields_projects, :id, :primary_key)
  TRANSACTION (0.8ms)  BEGIN
   (12.7ms)  ALTER TABLE "custom_fields_projects" ADD "id" bigserial primary key
   -> 0.0132s
-- execute("WITH cte AS ( SELECT row_number() OVER () AS row_num, project_id, custom_field_id FROM custom_fields_projects ) UPDATE custom_fields_projects SET id = cte.row_num FROM cte WHERE custom_fields_projects.project_id = cte.project_id AND custom_fields_projects.custom_field_id = cte.custom_field_id;")
   (1.1ms)  WITH cte AS ( SELECT row_number() OVER () AS row_num, project_id, custom_field_id FROM custom_fields_projects ) UPDATE custom_fields_projects SET id = cte.row_num FROM cte WHERE custom_fields_projects.project_id = cte.project_id AND custom_fields_projects.custom_field_id = cte.custom_field_id;
   -> 0.0015s
== 20240917105829 AddPrimaryKeyToCustomFieldsProjects: migrated (0.0150s) =====
```

```sql
                                Table "public.custom_fields_projects"
     Column      |  Type  | Collation | Nullable |                      Default
-----------------+--------+-----------+----------+----------------------------------------------------
 custom_field_id | bigint |           | not null |
 project_id      | bigint |           | not null |
 id              | bigint |           | not null | nextval('custom_fields_projects_id_seq'::regclass)
Indexes:
    "custom_fields_projects_pkey" PRIMARY KEY, btree (id)
    "index_custom_fields_projects_on_custom_field_id_and_project_id" UNIQUE, btree (custom_field_id, project_id)
Foreign-key constraints:
    "fk_rails_12fb30588e" FOREIGN KEY (project_id) REFERENCES projects(id) ON UPDATE CASCADE ON DELETE CASCADE
    "fk_rails_e51cefe60d" FOREIGN KEY (custom_field_id) REFERENCES custom_fields(id) ON UPDATE CASCADE ON DELETE CASCADE
```

cc @ulferts @machisuji

akabiru replied:
https://github.com/opf/openproject/pull/16694/commits/922bd158ef38b08acffa45737b0f4cab4b219e50

> _An ActiveRecord primary key migration translates to a BIGSERIAL PRIMARY KEY
constraint in Postgres. BigSerial is an auto-incrementing numeric type so there
is no need to manually backfill_

akabiru committed:
chore[Op#57516]: remove unnecessary manual backfilling

An ActiveRecord primary key migration translates to a BIGSERIAL PRIMARY KEY
constraint in Postgres. BigSerial is an auto-incrementing numeric type so there
is no need to manually backfill
